overlord/hookstate/hookmgr.go

```diff
@@ -0,0 +1,122 @@
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package hookstate implements the manager and state aspects responsible for
+// the running of hooks.
+package hookstate
+
+import (
+	"fmt"
+	"regexp"
+
+	"gopkg.in/tomb.v2"
+
+	"github.com/snapcore/snapd/overlord/state"
+)
+
+// HookManager is responsible for the maintenance of hooks in the system state.
+// It runs hooks when they're requested, assuming they're present in the given
+// snap. Otherwise they're skipped with no error.
+type HookManager struct {
+	state      *state.State
+	runner     *state.TaskRunner
+	repository *repository
+}
+
+type Handler interface {
+	Before() error
+	Done() error
+	Error(err error) error
+}
+
+type HandlerGenerator func(*Context) Handler
+
+// Manager returns a new HookManager.
+func Manager(s *state.State) (*HookManager, error) {
+	runner := state.NewTaskRunner(s)
+	manager := &HookManager{
+		state:      s,
+		runner:     runner,
+		repository: newRepository(),
+	}
+
+	runner.AddHandler("run-hook", manager.doRunHook, nil)
+
+	return manager, nil
+}
+
+// Register requests that a given handler generator be called when a matching
+// hook is run, and the handler be used for the hook.
+//
+// Specifically, if a matching hook is about to be run, the handler's Before()
+// method will be called. After the hook has completed running, either the
+// handler's Done() method or its Error() method will be called, depending on
+// the outcome.
+func (m *HookManager) Register(pattern *regexp.Regexp, generator HandlerGenerator) {
+	m.repository.addHandlerGenerator(pattern, generator)
+}
+
+// Ensure implements StateManager.Ensure.
+func (m *HookManager) Ensure() error {
+	m.runner.Ensure()
+	return nil
+}
+
+// Wait implements StateManager.Wait.
+func (m *HookManager) Wait() {
+	m.runner.Wait()
+}
+
+// Stop implements StateManager.Stop.
+func (m *HookManager) Stop() {
+	m.runner.Stop()
+}
+
+// doRunHook actually runs the hook that was requested.
+//
+// Note that this method is synchronous, as the task is already running in a
+// goroutine.
+func (m *HookManager) doRunHook(task *state.Task, tomb *tomb.Tomb) error {
+	task.State().Lock()
+	var hook hookSetup
+	err := task.Get("hook-setup", &hook)
+	task.State().Unlock()
+
+	if err != nil {
+		return fmt.Errorf("failed to extract hook from task: %s", err)
+	}
+
+	// Obtain a list of handlers for this hook (if any)
+	handlers := m.repository.generateHandlers(newContext(task, hook))
```

Comment:
handler being empty is probably an error or we should do nothing (the latter is more the task runner behavior atm), see also my point below about multiple handlers